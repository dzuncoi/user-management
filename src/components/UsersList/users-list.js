import React from 'react'
import PropTypes from 'prop-types'
import { Button, Table, Tooltip } from 'antd'
import { Link } from 'react-router-dom'

import Panel from '../Panel/panel'

import {
  NOT_LOADED,
  LOADING,
  SUCCEEDED,
  FAILED,
} from '../../constants/loading-status'

export const columns = [{
  title: 'Full Name',
  key: 'fullName',
  render: (val, record) => <Link to={`user/${record.id}`}>{record.firstName} {record.lastName}</Link>,
}, {
  title: 'Email',
  key: 'email',
  dataIndex: 'email',
}, {
  title: 'Groups',
  key: 'groups',
  render: (val, { groups = [] }) => (groups.length <= 1
    ? groups[0].name
    : (
      <Tooltip
        title={
          <React.Fragment>
            {groups.map(group => <div key={group.id}>{group.name}</div>)}
          </React.Fragment>
        }
      >
        {`${groups.length} groups`}
      </Tooltip>
    )
  ),
}, {
  title: 'Actions',
  key: 'actions',
  render: () => <Button type="ghost">Delete</Button>,
  fixed: 'right',
  width: 100,
}]

export const UsersListComp = props => (
  <Panel
    title="Users"
    extra={
      <Link to="user/new">
        <Button type="primary">New User</Button>
      </Link>
    }
  >
    <Table
      name="users"
      rowKey="id"
      dataSource={props.data}
      columns={columns}
      loading={props.loadingStatus === LOADING}
      scroll={{ x: true }}
    />
  </Panel>
)

UsersListComp.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    email: PropTypes.string,
  })),
  loadingStatus: PropTypes.oneOf([NOT_LOADED, LOADING, SUCCEEDED, FAILED]),
}

UsersListComp.defaultProps = {
  data: [],
  loadingStatus: NOT_LOADED,
}

export default React.memo(UsersListComp)
