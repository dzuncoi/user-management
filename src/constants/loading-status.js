export const LOADING = 'loading'
export const SUCCEEDED = 'succeeded'
export const FAILED = 'failed'
export const NOT_LOADED = 'not_loaded'
