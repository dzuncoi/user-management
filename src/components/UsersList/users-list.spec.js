import React from 'react'

import { UsersListComp } from './users-list'
import { LOADING, NOT_LOADED } from '../../constants/loading-status'

jest.mock('antd', () => ({
  Button: () => 'Button',
  Table: () => 'Table',
  Tooltip: () => 'Tooltip',
}))
jest.mock('react-router-dom', () => ({
  Link: () => 'Link',
}))
jest.mock('../Panel/panel', () => 'Panel')

describe('users-list', () => {
  let props

  beforeEach(() => {
    props = {
      loadingStatus: NOT_LOADED,
      data: [{
        id: '1',
        name: 'data-1',
      }, {
        id: '2',
        name: 'data-2',
      }],
      deleteUser: jest.fn(),
    }
  })

  it('renders', () => {
    expect(shallow(<UsersListComp {...props} />)).toMatchSnapshot()
  })

  it('renders loading on table if loadingStatus is LOADING', () => {
    const wrapper = shallow(<UsersListComp {...props} loadingStatus={LOADING} />)
    expect(wrapper.find('Table').props().loading).toBeTruthy()
  })

  describe('table columns', () => {
    let cols

    beforeEach(() => {
      const w = shallow(<UsersListComp {...props} />)
      cols = w.instance().columns
    })

    it('renders fullName with Link to edit user', () => {
      const Col = cols[0].render(null, {
        id: 'id',
        firstName: 'first name',
        lastName: 'last name',
      })
      expect(Col.props.to).toEqual('user/id')
      expect(Col.props.children.join('')).toEqual('first name last name')
    })

    it('renders group name if user belongs to only 1 group', () => {
      const Col = cols[2].render(null, {
        groups: [{ name: 'unique group' }],
      })
      expect(Col).toEqual('unique group')
    })

    it('renders amount of groups if user belongs to multiple groups', () => {
      const Col = cols[2].render(null, {
        groups: [{ name: 'group 1', id: '1' }, { name: 'group 2', id: '2' }],
      })
      expect(Col.props.children).toEqual('2 groups')
    })
  })
})
