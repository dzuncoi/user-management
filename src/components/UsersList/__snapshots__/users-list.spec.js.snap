// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users-list renders 1`] = `
<Panel
  extra={
    <Link
      to="user/new"
    >
      <Button
        type="primary"
      >
        New User
      </Button>
    </Link>
  }
  title="Users"
>
  <Table
    columns={
      Array [
        Object {
          "key": "fullName",
          "render": [Function],
          "title": "Full Name",
        },
        Object {
          "dataIndex": "email",
          "key": "email",
          "title": "Email",
        },
        Object {
          "key": "groups",
          "render": [Function],
          "title": "Groups",
        },
        Object {
          "fixed": "right",
          "key": "actions",
          "render": [Function],
          "title": "Actions",
          "width": 100,
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "id": "1",
          "name": "data-1",
        },
        Object {
          "id": "2",
          "name": "data-2",
        },
      ]
    }
    loading={false}
    name="users"
    rowKey="id"
    scroll={
      Object {
        "x": true,
      }
    }
  />
</Panel>
`;
